client/src/loaders/JsonLoader.js

```diff
@@ -13,12 +13,21 @@ class JsonLoader extends Component {
     this.fetchJson = this.fetchJson.bind(this);
   }
 
+  //TODO: JSON is sent and recieved now. Not sure what happens after that
   fetchJson() {
     const {path, query, options} = this.props;
     const url = (_.isEmpty(query))
       ? path
       : `${path}?${qs.stringify(query)}`;
-    return fetch(url, options).then(r => r.json());
+    return fetch(url, options)
+      .then(r => {
+        if (r.status === 200){
+          return r.json();
+        } else{
+          return {};
+        }
+      })
+      .catch();
```

Comment:
Could we remove this and let callers deal with handling the response to `fetch`, or is this important to make `JsonLoader` work as it should?

Reply:
I'm not sure what you mean by letting the callers handle this but I'd be happy to hear more.

the return {} is needed because otherwise fetching with the wrong credentials returns undefined which breaks the if/else check in PromiseLoader to indicate if the promise was resolved or rejected. 

The .catch() was used to handle a "Unexpected end of JSON input" error. I can't seem to reproduce the problem now but I can look deeper into it when testing. Maybe the fix you have in mind will give a new reason to get rid of it too

Reply:
@keving17 This is a great explanation, thanks! 👍 

So there are a few pieces here that are interrelated, and I think I'm following what you were seeing and how you went above resolving it with the changes to `JsonLoader` and `PromiseLoader`.

1. First, let's look at `JsonLoader`.  When the wrong credentials returned, and the status code wasn't 200, and the response body wasn't parseable, that led to an error.  You worked around this by only parsing as JSON if there was a 200, and returning `{}` if not.

But in the case where the server returns a different status code, or the JSON isn't parseable, I think we want that to reject the promise instead (in other words, if `JsonLoader` is supposed to load some JSON from the server, and it gets back an unexpected status code from the server, or the server doesn't return actual JSON, this should be considered an error path not a success path.

So here that would mean changing the `else` branch so that it does something like `return Promise.reject({ status: r.status })` or `return Promise.reject({ error: "Could not parse JSON response" })`

2. With this change, then I think we don't need the changes you had to work out in `PromiseLoader`, and I'll comment over on that code about that. :)

Reply:
FWIW, what I'm suggesting in #1 above is the same way you handled the then/catch cases in `authenticate`: https://github.com/mit-teaching-systems-lab/swipe-right-for-cs/pull/93/files#diff-1af36ee804d1296b95a69a669bb5fb03R51

Reply:
Changed code to follow your suggestion. Everything seems to work fine 👍